docs/firebase-ui-firestore.md

```diff
@@ -0,0 +1,217 @@
+
+```dart
+FirestoreListView<Map<String, dynamic>>(
+  loadingBuilder: (context) => MyCustomLoadingIndicator(),
+  errorBuilder: (context, error, stackTrace) => MyCustomError(error, stackTrace),
+  // ...
+);
+```
+
+### Advanced configuration
+
+In many cases, the `FirestoreListView` widget is enough to render simple lists of collection data.
+However, you may have specific requirements which require more control over the widget's behavior
+(such as using a [`GridView`](https://api.flutter.dev/flutter/widgets/GridView-class.html)).
+
+The `FirestoreQueryBuilder` provides the building blocks for advanced configuration at the expense of
+requiring more boilerplate code. The widget does not provide any underlying list implementation, instead
+you are expected to provide this yourself.
+
+Much like the `FirestoreListView` widget, provide a query and builder:
+
+```dart
+final usersQuery = FirebaseFirestore.instance.collection('users').orderBy('name');
+
+FirestoreQueryBuilder<Map<String, dynamic>>(
+  query: usersQuery,
+  builder: (context, snapshot, _) {
+    // ... TODO!
+  },
+);
+```
+
+The main difference to note here is that the `builder` property returns a `QueryBuilderSnapshot`, rather
+than an individual document. The builder returns the current state of the entire query, such as whether
+data is loading, an error has occurred and the documents.
+
+This requires us to implement our own list based implementation. Firstly, let's handle the loading and error
+states:
+
+```dart
+FirestoreQueryBuilder<Map<String, dynamic>>(
+  query: usersQuery,
+  builder: (context, snapshot, _) {
+    if (snapshot.isFetching) {
+      return const CircularProgressIndicator();
+    }
+
+    if (snapshot.hasError) {
+      return Text('Something went wrong! ${snapshot.error}');
+    }
+
+    // ...
+  },
+);
+```
+
+Next, we now need to return a list-view based implementation for our application to display the data. For example,
+to display a grid of users, we can use the [`GridView`](https://api.flutter.dev/flutter/widgets/GridView-class.html) widget:
+
+```dart
+FirestoreQueryBuilder<Map<String, dynamic>>(
+  query: usersQuery,
+  builder: (context, snapshot, _) {
+    // ...
+
+    return GridView.builder(
+      itemCount: snapshot.docs.length,
+      itemBuilder: (context, index) {
+        // if we reached the end of the currently obtained items, we try to
+        // obtain more items
+        if (snapshot.hasMore && index + 1 == snapshot.docs.length) {
+          // Tell FirestoreQueryBuilder to try to obtain more items.
+          // It is safe to call this function from within the build method.
+          snapshot.fetchMore();
+        }
+
+        final user = snapshot.docs[index].data();
+
+        return Container(
+          padding: const EdgeInsets.all(8),
+          color: Colors.teal[100],
+          child: const Text("User name is ${user['name']}"),
+        );
+      },
+    );
+  },
+);
+```
+
+With more power comes more responsibility:
```

Comment:
I doubt they'll allow this on Firebase docs. I'd make it a sub-header with something along the lines of "manually fetching data within `GridView` implementation"